Web Extensions: Add getSubTree(), getChildren(), and get()

akutira-umich opened the pull request:
#### afff0029f78abd5b17a5b8bf581427087b0ad5d1
<pre>
Web Extensions: Add getSubTree(), getChildren(), and get()
<a href="https://rdar.apple.com/156613332">rdar://156613332</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296436">https://bugs.webkit.org/show_bug.cgi?id=296436</a>

Reviewed by Timothy Hatcher.

Adds the WebProcess and UIProcess for get(), getSubTree(), and getChildren(). Also adds tests for these.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::findBookmarkNodeInTree):
(WebKit::createShallowParametersFromProtocolObjects):
(WebKit::WebExtensionContext::bookmarksGetSubTree):
(WebKit::WebExtensionContext::bookmarksGet):
(WebKit::WebExtensionContext::bookmarksGetChildren):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::getChildren):
(WebKit::WebExtensionAPIBookmarks::getSubTree):
(WebKit::WebExtensionAPIBookmarks::get):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode bookmarkTypeForWebExtensionContext:]):
(-[_MockBookmarkNode childrenForWebExtensionContext:]):
(-[TestBookmarksDelegate _webExtensionController:createBookmarkWithParentIdentifier:index:url:title:forExtensionContext:completionHandler:]):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, GetSubTree)):

Canonical link: <a href="https://commits.webkit.org/298215@main">https://commits.webkit.org/298215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f1cb24456592ad17356150f66572dcacc4a5745

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114212 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33959 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120378 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64948 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34588 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42521 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86796 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41747 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117160 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27540 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67185 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26723 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20703 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96913 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123595 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41231 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30740 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95625 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41608 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98779 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95408 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40553 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18376 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37319 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18371 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41109 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46621 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40717 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44022 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -261,7 +261,24 @@
-    callback->reportError(@"unimplemented");
+    WebProcess::singleton().sendWithAsyncReply(
+        Messages::WebExtensionContext::BookmarksGetChildren(bookmarkIdentifier),
+        [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<Vector<WebExtensionBookmarksParameters>, WebExtensionError>&& result) {
+            if (!result) {
+                callback->reportError(result.error().createNSString().get());
+                return;
+            }
+            const Vector<WebExtensionBookmarksParameters>& resultVector = result.value();
+
+            NSMutableArray *resultArray = [NSMutableArray arrayWithCapacity:resultVector.size()];
+            for (const auto& node : resultVector)
+                [resultArray addObject:toAPI(node)];
+
+            callback->call(resultArray);
+        },
+
+        extensionContext().identifier()
```
This is usually best to break out into a separate lambda, so it is easier to read.

akutira-umich replied:
I looked at other API's like Scripting and SidePanel, the format seems to be to keep it inline with the lambda.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -294,7 +311,30 @@
-    callback->reportError(@"unimplemented");
+    WebProcess::singleton().sendWithAsyncReply(
+        Messages::WebExtensionContext::BookmarksGetSubTree(bookmarkIdentifier),
+        [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<Vector<WebExtensionBookmarksParameters>, WebExtensionError>&& result) {
+            if (!result) {
+                callback->reportError(result.error().createNSString().get());
+                return;
+            }
+
+            const Vector<WebExtensionBookmarksParameters>& resultVector = result.value();
+            const WebExtensionBookmarksParameters& subtreeRootNode = resultVector[0];
+            if (!subtreeRootNode.children) {
+                callback->call(@[], nil);
+                return;
+            }
+
+            NSMutableArray *childrenArray = [NSMutableArray arrayWithCapacity:subtreeRootNode.children->size()];
+            for (const auto& childNode : *subtreeRootNode.children)
+                [childrenArray addObject:toAPI(childNode)];
+
+            callback->call(childrenArray);
+        },
```
Ditto.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -317,7 +357,42 @@
+
+    WebProcess::singleton().sendWithAsyncReply(
+        Messages::WebExtensionContext::BookmarksGet(bookmarkIds),
+        [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<Vector<WebExtensionBookmarksParameters>, WebExtensionError>&& result) {
+            if (!result) {
+                callback->reportError(result.error().createNSString().get());
+                return;
+            }
+
+            const Vector<WebExtensionBookmarksParameters>& resultVector = result.value();
+
+            NSMutableArray *resultArray = [NSMutableArray arrayWithCapacity:resultVector.size()];
+            for (const auto& node : resultVector)
+                [resultArray addObject:toAPI(node)];
+
+            callback->call(resultArray);
+        },
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -75,8 +75,17 @@ - (NSInteger)indexForWebExtensionContext:(WKWebExtensionContext *)context
+    if (url && url.length > 0)
+        return _WKWebExtensionBookmarkTypeBookmark;
+    return _WKWebExtensionBookmarkTypeFolder;
+}
+- (NSArray<id<_WKWebExtensionBookmark>> *)childrenForWebExtensionContext:(WKWebExtensionContext *)context {
```
```suggestion
}

- (NSArray<id<_WKWebExtensionBookmark>> *)childrenForWebExtensionContext:(WKWebExtensionContext *)context {
```

akutira-umich committed:
Web Extensions: Add getSubTree(), getChildren(), and get()
rdar://156613332
https://bugs.webkit.org/show_bug.cgi?id=296436

Reviewed by Timothy Hatcher.

Adds the WebProcess and UIProcess for get(), getSubTree(), and getChildren(). Also adds tests for these.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::findBookmarkNodeInTree):
(WebKit::createShallowParametersFromProtocolObjects):
(WebKit::WebExtensionContext::bookmarksGetSubTree):
(WebKit::WebExtensionContext::bookmarksGet):
(WebKit::WebExtensionContext::bookmarksGetChildren):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::getChildren):
(WebKit::WebExtensionAPIBookmarks::getSubTree):
(WebKit::WebExtensionAPIBookmarks::get):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode bookmarkTypeForWebExtensionContext:]):
(-[_MockBookmarkNode childrenForWebExtensionContext:]):
(-[TestBookmarksDelegate _webExtensionController:createBookmarkWithParentIdentifier:index:url:title:forExtensionContext:completionHandler:]):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, GetSubTree)):

Canonical link: https://commits.webkit.org/298215@main
